content/en/BestPractices/InputValidation.md

```diff
@@ -0,0 +1,115 @@
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+### Analysis Of The Issue
+
+Here, the [quantityCheck() function](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L590) is called inside a middleware function [mounted on the /api/BasketItems API path](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370). It checks the quantity in the request payload against some validation logic. However, the validation logic is flawed. It only checks if the quantity exceeds the limit per user and the limit per order (and skips the check if it is a deluxe user). But, it doesn’t check if the quantity is positive.
+
+This small mistake turns out to be very costly. Users can freely input negative values in the `quantity` parameter. An attacker can abuse this by adding a negative quantity of the item (say, `-100`) to their basket. The basket value gets deducted from their wallet balance leading to a net positive amount in the attacker’s wallet. In this way, the attacker earns free money equivalent to `100` times the cost of the item.
+
+### What Is The Impact?
+An attacker might execute this process repeatedly. Without timely detection, it might lead to substantial business losses. Thus, it can be a **very high-impact** issue.
+
+### How Could This Be Mitigated? 
+1. Check if the quantity is greater than zero (`quantity > 0`),
+2. Also, perform decimal integer check with `Number.isInteger(quantity, 10)`. It returns `true` if the value is a decimal integer and `false` otherwise,
+3. If the above checks fail, send a `400` error response.
+
+```
+async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
+  const product = await QuantityModel.findOne({ where: { ProductId: id } })
+  if (!product) {
+    throw new Error('No such product found!')
+  }
+
+ // Check if the quantity is a positive decimal integer value, if not send a `400` error response
+ if(!Number.isInteger(quantity, 10) || quantity <= 0) {
+      res.status(400).json({ error: res.__(‘Invalid Quantity of items.') })
+ }
+
+  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
```

Comment:
The main thing I would suggest is perhaps adding other examples or types of possible issues. We focus on this one Business Logic example, which I think illustrates the point nicely. I would suggest also talking about SQL Injection (SQLi) or HTML Injection, or other types of vulnerabilities related to Input Validation.

I think at least a 1-2 sentence description of some other scenarios would drive home that it's not *just* business logic we're solving with this, and prompt readers to look for other areas where user-supplied input could be dangerous. 

As a thought, is it also worth mentioning Output Encoding? If not in detail, then as a space where we can cross-reference with another article about Encoding. (To be honest, I'm mostly asking because I see Input Validation and Output Encoding recommended together so often, it seems weird to see one without the other).

Reply:
Thanks. I wanted to put XSS and SQLi. But unfortunately, input validation alone can't fix them, for example in SQLi, the problem can be solved through parameterized queries and the use of safe ORM functions. XSS could also be fixed with encoding input to entities, depending on the context. 

That is a good idea, I can cross reference the output encoding article.